### PR TITLE
Differential expression performance improvements

### PR DESCRIPTION
**Pull request by blrnw3**
#### Reviewers

**Functional:**   @atolopko-czi  
**Readability:**   @ebezzi 

---

## Changes
* Remove legacy h5ad diffex algo
* Update to latest tiledb
* Optimize cxg diffex algo
   * remove parallelism logic (tdb handles this better) and config params for this code
   * tweak tdb queries based on Bruce's research
* Optimize other tdb queries (the get_X_array calls)
* Remove col shift logic - legacy
* Update tdb context config and related tests
* Update run_diffex script

### CODE REVIEW

File: server/compute/diffexp_cxg.py [hunk 1]
```diff
@@ -1,120 +1,65 @@
...
+    :param maskB: observation selection mask for set 2
+    :param top_n: number of variables to return stats for
+    :param diffexp_lfc_cutoff: minimum
+    :param arr: the tdb array to read
+    absolute value returning [ varindex, logfoldchange, pval, pval_adj ] for top N genes
+    :return:  for top N genes, {"positive": for top N genes, [ varindex, foldchange, pval, pval_adj ],
+              "negative": for top N genes, [ varindex, foldchange, pval, pval_adj ]}
+    """
+    matrix = adaptor.open_array(arr)
     row_selector_A = np.where(maskA)[0]
```
Comment (bkmartinjr) [inline server/compute/diffexp_cxg.py:45]: Using `np.where` is slightly poor form and lacks clarity.  See numpy [docs](https://numpy.org/doc/stable/reference/generated/numpy.where.html):

> When only condition is provided, this function is a shorthand for np.asarray(condition).[nonzero](https://numpy.org/doc/stable/reference/generated/numpy.nonzero.html#numpy.nonzero)(). Using nonzero directly should be preferred, as it behaves correctly for subclasses.

Comment (bkmartinjr) [reply server/compute/diffexp_cxg.py:45]: Oh, and I apologize if this was originally my error!

File: server/dataset/cxg_dataset.py [hunk 2]
```diff
@@ -23,7 +23,12 @@
 class CxgDataset(Dataset):
     # TODO:  The tiledb context parameters should be a configuration option
     tiledb_ctx = tiledb.Ctx(
-        {"sm.tile_cache_size": 8 * 1024 * 1024 * 1024, "sm.num_reader_threads": 32, "vfs.s3.region": "us-east-1"}
+        {
+            "sm.tile_cache_size": 16 * 1024 ** 3,
```
Comment (bkmartinjr) [inline server/dataset/cxg_dataset.py:27]: will these changes be accompanied with a change of instance type?  The large tile_cache_size and py.init_buffer_bytes are suitable for a high-mem instance, but not for the current 64GB instances.  Thought: should these be pulled out into a deployment configuration (eg, set via env vars), or made adaptive (eg, tile cache is X% of RAM on the instance)?

File: server/compute/diffexp_cxg.py [hunk 3]
```diff
@@ -1,120 +1,65 @@
...
-    if diffexp_thread_executor is None:
-        diffexp_thread_executor = concurrent.futures.ThreadPoolExecutor(max_workers=max_workers)
-    return diffexp_thread_executor
-
-
-def diffexp_ttest(adaptor, maskA, maskB, top_n=8, diffexp_lfc_cutoff=0.01):
-
-    matrix = adaptor.open_array("X")
+def set_config(*args, **kwargs):
+    # TODO: find callers and eliminate calls
```
[the remark this answers is not in the document]
Comment (blrnw3) [reply server/compute/diffexp_cxg.py:12]: I tried, but there are indeed callers, didn't track them down but we should

File: server/compute/diffexp_cxg.py [hunk 4]
```diff
@@ -165,35 +199,14 @@ def _mean_var_sparse_numba(x, var, nrows, ncols):
...
-    if fp_err_occurred:
-        mean[np.isfinite(mean) == False] = 0  # noqa: E712
-        v[np.isfinite(v) == False] = 0  # noqa: E712
-    else:
-        mean[np.isnan(mean)] = 0
-        v[np.isnan(v)] = 0
-
-    return mean, v
+def mean_var_cnt_dense(matrix, _, rows):
+    # TODO: why is the list cast still necessary? Tdb is throwing an error
```
[the remark this answers is not in the document]
Comment (blrnw3) [reply server/compute/diffexp_cxg.py:210]: Agreed, will add to the task list

File: server/compute/diffexp_cxg.py (hunk 1 above)
Comment (blrnw3) [reply server/compute/diffexp_cxg.py:45]: This was the original code, not sure who wrote it but good point, we should update it!

File: server/dataset/cxg_dataset.py (hunk 2 above)
Comment (blrnw3) [reply server/dataset/cxg_dataset.py:27]: Yes there's an infra PR here: https://github.com/chanzuckerberg/single-cell-infra/pull/526
Re: making them dynamic based on instance - agreed!

File: server/dataset/cxg_dataset.py [hunk 5]
```diff
@@ -265,18 +270,12 @@ def get_X_array(self, obs_mask=None, var_mask=None):
             if obs_items == slice(None) and var_items == slice(None):
                 data = X[:, :]
```
[the remark this answers is not in the document]
Comment (blrnw3) [reply server/dataset/cxg_dataset.py:271]: Good spot. Though is this even used. TODO: check if it is

File: server/tests/unit/dataset/test_cxg_dataset.py [hunk 6]
```diff
@@ -15,6 +20,15 @@ def test_get_colors(self):
         self.assertDictEqual(data.get_colors(), dict())
 
     def get_data(self, fixture):
+        # FIXTURES_ROOT = "/home/ec2-user/server/tests/fixtures"
         data_locator = f"{FIXTURES_ROOT}/{fixture}"
         config = app_config(data_locator)
         return CxgDataset(DataLocator(data_locator), config)
+
+    def test_tdb_bug(self):
+        """ This gives different results on 0.12.4 vs 0.13.1. TODO: fix """
```
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply server/tests/unit/dataset/test_cxg_dataset.py:29]: resolved by work-around added to server/dataset/cxg_dataset.py:set_tiledb_context()

File: server/compute/diffexp_cxg.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply server/compute/diffexp_cxg.py:210]: Update: this is a bug in recent TileDB-Py versions. In the queue for a fix in TileDB 0.13.2, at which point multi_index should again support either list or ndarray.

File: server/compute/diffexp_cxg.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply server/compute/diffexp_cxg.py]: tracked down and removed.

File: server/compute/diffexp_cxg.py (hunk 1 above)
Comment (bkmartinjr) [reply server/compute/diffexp_cxg.py]: modified to us numpy recommended style

File: server/dataset/cxg_dataset.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply server/dataset/cxg_dataset.py:278]: It is used.  Added a commit to simplify the logic, as multi_index[] can provide both query types -- we don't need the branch.

**Review by bkmartinjr — APPROVED** [review]
I have resolved my own prior comments, but would like additional reviewers for the new modifications.

File: server/dataset/cxg_dataset.py [hunk 7]
```diff
@@ -21,9 +21,13 @@
...
 
 class CxgDataset(Dataset):
-    # TODO:  The tiledb context parameters should be a configuration option
+    # These defaults are overridden by the config variable: server.adaptor.cxg_adaptor.tiledb_cxt
     tiledb_ctx = tiledb.Ctx(
-        {"sm.tile_cache_size": 8 * 1024 * 1024 * 1024, "sm.num_reader_threads": 32, "vfs.s3.region": "us-east-1"}
+        {
+            "sm.tile_cache_size": 8 * 1024 ** 3,
+            "py.init_buffer_bytes": 16 * 1024 ** 3,
+            "vfs.s3.region": "us-east-1",
```
Comment (ebezzi) [inline server/dataset/cxg_dataset.py:29]: I was wondering why the region is even required. Our buckets live in `us-west-2` so it looks like this is ignored by TileDB.

File: server/dataset/cxg_dataset.py [hunk 8]
```diff
@@ -244,7 +240,7 @@ def __remap_indices(self, coord_range, coord_mask, coord_data):
         if coord_mask is None:
             return coord_range, coord_data
 
-        indices = np.where(coord_mask)[0]
+        indices = coord_mask.nonzero()[0]
```
Comment (ebezzi) [inline server/dataset/cxg_dataset.py:243]: Out of curiosity, is this more efficient than using `np.where`?

**Review by ebezzi — APPROVED** [review]

File: server/dataset/cxg_dataset.py (hunk 8 above)
Comment (bkmartinjr) [reply server/dataset/cxg_dataset.py:243]: no.  They are literally identical, but np.where is deprecated in this case.  See the numpy documentation here: https://numpy.org/doc/stable/reference/generated/numpy.where.html

It says:
```
Note

When only condition is provided, this function is a shorthand for np.asarray(condition).[nonzero](https://numpy.org/doc/stable/reference/generated/numpy.nonzero.html#numpy.nonzero)(). Using nonzero directly should be preferred, as it behaves correctly for subclasses. The rest of this documentation covers only the case where all three arguments are provided.
```

File: server/dataset/cxg_dataset.py (hunk 7 above)
Comment (bkmartinjr) [reply server/dataset/cxg_dataset.py:29]: The idea was just to have an explicit default in the code so we didn't have to rely on a TileDB default.  You are right that the current configurations override it to match our current deployment.  As or deployment evolves (eg, multi-region?), the config could change.

**Review by atolopko-czi — APPROVED** [review]
Approving, with a few minor comments!

**Pull request merged**